os/include/tinyara/mipidsi/mipi_dsi.h

```diff
@@ -0,0 +1,907 @@
+                                                  * transfer done */
+#define MIPI_DSI_HS_PKT_END_ALIGNED    (1 << 13) /* Transmit data ending at
+                                                  * the same time for all
+                                                  * lanes within one hsync */
+
+/****************************************************************************
+ * Public Types
+ ****************************************************************************/
+
+struct mipi_dsi_host;
+struct mipi_dsi_device;
+
+/* Read/write DSI buffer */
+
+struct mipi_dsi_msg
+{
+  uint8_t channel;          /* Virtual channel id */
+  uint8_t type;             /* Payload data type */
+  uint8_t flags;            /* Flags controlling this message transmission */
+
+  size_t tx_len;            /* Length of tx_buf */
+  FAR const void *tx_buf;   /* Data to be written */
+
+  size_t rx_len;            /* Lenght of rx_buf */
+  FAR void *rx_buf;         /* Data to be read, or NULL */
+};
+
+/* Represents a MIPI DSI packet in protocol format */
+
+struct mipi_dsi_packet
+{
+  size_t size;                 /* Size (in bytes) of the packet */
+  uint8_t header[4];           /* The four bytes that make up the header
+                                * (Data ID, Word Count or Packet Data,
+                                * and ECC) */
+  size_t payload_length;       /* Number of bytes in the payload */
+  FAR const uint8_t *payload;  /* A pointer to a buffer containing
+                                * the payload */
+};
+
+/* DSI bus operations
+ *
+ * DSI packets transmitted by .transfer() are passed in as mipi_dsi_msg
+ * structures. This structure contains information about the type of packet
+ * being transmitted as well as the transmit and receive buffers. When an
+ * error is encountered during transmission, this function will return a
+ * negative error code. On success it shall return the number of bytes
+ * transmitted for write packets or the number of bytes received for read
+ * packets.
+ *
+ * Note that typically DSI packet transmission is atomic, so the .transfer()
+ * function will seldomly return anything other than the number of bytes
+ * contained in the transmit buffer on success.
+ *
+ * Also note that those callbacks can be called no matter the state the
+ * host is in. Drivers that need the underlying device to be powered to
+ * perform these operations will first need to make sure it's been
+ * properly enabled.
+ */
+
+struct mipi_dsi_host_ops
+{
+  /* Attach DSI device to DSI host */
+
+  CODE int (*attach)(FAR struct mipi_dsi_host *host,
+                     FAR struct mipi_dsi_device *device);
+
+  /* Detach DSI device from DSI host */
+
+  CODE int (*detach)(FAR struct mipi_dsi_host *host,
+                     FAR struct mipi_dsi_device *device);
+
+  /* Transmit a DSI packet */
+
+  CODE ssize_t (*transfer)(FAR struct mipi_dsi_host *host,
+                           FAR const struct mipi_dsi_msg *msg);
+};
```

Comment:
There are many coding rules violations. Please check them all once.
Once we modify them, let's merge the PR.